colcon_core/verb/__init__.py

```diff
@@ -124,6 +124,41 @@ def check_and_mark_install_layout(install_base, *, merge_install):
     marker_path.write_text(this_install_layout + '\n')
 
 
+def check_and_mark_root_dir(this_build_tool='colcon'):
+    """
+    Check the marker file for root workspace.
+
+    If marker file is found in parent directory, raise error;
+    if it is found in current direcotry, continue to build.
+    If no marker file found, create marker file.
+
+    The marker filename is `.colcon`.
+
+    :param str this_build_tool: The name of this build tool
+    :raises RuntimeError: if marker file is found in parent directory
+    """
+    original_path = Path(os.getcwd())
+    current_path = original_path
+    parent_path = current_path.parent
+    while current_path.name != parent_path.name:
+        marker_path = current_path / '.colcon'
+        if marker_path.parent.is_dir() and marker_path.is_file():
+            if current_path.name == original_path.name:
+                return
+            raise RuntimeError(
+                "'{original_path}' is not marked as the root directory. "
+                "Please go to '{current_path}' for `colcon build`. "
+                'If the current directory is intended to be the root '
+                "workspace, please remove the '.colcon' file "
+                "in '{current_path}'.".format_map(locals()))
+        else:
+            current_path = parent_path
+            parent_path = current_path.parent
+
+    marker_path = original_path / '.colcon'
```

Comment:
A more expressive filename would be helpful to convey the semantic meaning of the marker file.